File: src/memento/repository/transactions.py
from __future__ import annotations

import sqlite3
import threading
from collections.abc import Callable
from dataclasses import dataclass
from pathlib import Path
from typing import Protocol

from memento.control.checkpoints import CheckpointError, CheckpointHook
from memento.control.operations import (
    OperationRecord,
    OperationRequest,
    OperationState,
    create_operation,
    list_interrupted_operations,
    mark_operation_conflict,
    mark_operation_failed,
    mark_operation_running,
    mark_operation_succeeded,
)
from memento.repository.git import (
    GitRepositoryPaths,
    commit_exact_paths,
    create_operation_worktree,
    diff_main_paths,
    exact_staged_paths,
    get_main_revision,
    materialize_current_checkout,
    publish_main_compare_and_swap,
    remove_operation_worktree,
    resolve_worktree_revision,
)

_TRANSACTION_LOCKS: dict[str, threading.Lock] = {}
_TRANSACTION_LOCKS_GUARD = threading.Lock()


def _transaction_lock(paths: GitRepositoryPaths) -> threading.Lock:
    key = str(paths.bare_dir.resolve())
    with _TRANSACTION_LOCKS_GUARD:
        return _TRANSACTION_LOCKS.setdefault(key, threading.Lock())


class TransactionConflictError(RuntimeError):
    """Raised when an optimistic concurrency check fails."""


MutationCallback = Callable[[Path], tuple[str, ...]]


class DerivedUpdateCallback(Protocol):
    def __call__(
        self,
        materialized_root: Path,
        repo_revision: str,
        changed_paths: tuple[str, ...],
    ) -> None: ...


@dataclass(frozen=True, slots=True)
class TransactionRequest:
    operation: OperationRequest
    expected_revision: str
    commit_message: str
    author_name: str
    author_email: str


@dataclass(frozen=True, slots=True)
class TransactionResult:
    operation: OperationRecord
    base_revision: str
    result_revision: str
    changed_paths: tuple[str, ...]
    materialized_path: Path
    replayed: bool = False


@dataclass(frozen=True, slots=True)
class RecoveryRecord:
    op_id: str
    state: OperationState
    classification: str
    revision: str | None


class TransactionManager:
    def __init__(
        self,
        connection: sqlite3.Connection,
        paths: GitRepositoryPaths,
        *,
        checkpoints: CheckpointHook | None = None,
        derived_update: DerivedUpdateCallback | None = None,
    ) -> None:
        self._connection = connection
        self._paths = paths
        self._checkpoints = checkpoints or CheckpointHook()
        self._derived_update = derived_update

    def apply(self, request: TransactionRequest, mutate: MutationCallback) -> TransactionResult:
        with _transaction_lock(self._paths):
            return self._apply_locked(request, mutate)

    def _apply_locked(
        self, request: TransactionRequest, mutate: MutationCallback
    ) -> TransactionResult:
        operation = create_operation(self._connection, request.operation)
        if operation.state is OperationState.SUCCEEDED and operation.result_revision is not None:
            payload = operation.replay_payload or {}
            replayed_paths = payload.get("changed_paths", ())
            if not isinstance(replayed_paths, list):
                replayed_paths = []
            return TransactionResult(
                operation=operation,
                base_revision=operation.base_revision or request.expected_revision,
                result_revision=operation.result_revision,
                changed_paths=tuple(str(path) for path in replayed_paths),
                materialized_path=self._paths.current_dir,
                replayed=True,
            )
        cleanup_worktree = True
        try:
            self._checkpoints.hit("operation_inserted")
            base_revision = get_main_revision(self._paths)
            if request.expected_revision != base_revision:
                conflict = mark_operation_conflict(
                    self._connection,
                    request.operation.op_id,
                    error_message=(
                        "expected revision "
                        f"{request.expected_revision} does not match {base_revision}"
                    ),
                )
                raise TransactionConflictError(conflict.error_message or "revision conflict")
            operation = mark_operation_running(
                self._connection,
                request.operation.op_id,
                base_revision=base_revision,
            )
            worktree = create_operation_worktree(
                self._paths,
                op_id=request.operation.op_id,
                base_revision=base_revision,
            )
            self._checkpoints.hit("worktree_created")
            changed_paths = tuple(sorted(mutate(worktree.path)))
            self._checkpoints.hit("mutation_applied")
            commit = commit_exact_paths(
                worktree,
                changed_paths=changed_paths,
                message=request.commit_message,
                author_name=request.author_name,
                author_email=request.author_email,
            )
            if exact_staged_paths(worktree.path):
                raise RuntimeError("staging area must be clean after commit")
            self._checkpoints.hit("commit_created")
            published = publish_main_compare_and_swap(
                self._paths,
                base_revision=base_revision,
                new_revision=commit.revision,
            )
            if not published:
                mark_operation_conflict(
                    self._connection,
                    request.operation.op_id,
                    error_message="repository head moved before publication",
                )
                raise TransactionConflictError("repository head moved before publication")
            self._checkpoints.hit("publication_complete")
            materialized = materialize_current_checkout(self._paths, revision=commit.revision)
            self._checkpoints.hit("current_materialized")
            if self._derived_update is not None:
                self._derived_update(materialized.path, commit.revision, commit.changed_paths)
            self._checkpoints.hit("derived_updated")
            operation = mark_operation_succeeded(
                self._connection,
                request.operation.op_id,
                result_revision=commit.revision,
                result={"changed_paths": list(commit.changed_paths)},
            )
            self._checkpoints.hit("operation_completed")
            return TransactionResult(
                operation=operation,
                base_revision=base_revision,
                result_revision=commit.revision,
                changed_paths=commit.changed_paths,
                materialized_path=materialized.path,
            )
        except TransactionConflictError:
            raise
        except CheckpointError:
            cleanup_worktree = False
            raise
        except Exception as exc:
            mark_operation_failed(
                self._connection,
                request.operation.op_id,
                error_class=exc.__class__.__name__,
                error_message=str(exc),
            )
            raise
        finally:
            if cleanup_worktree:
                remove_operation_worktree(self._paths, request.operation.op_id)

    def recover_startup(self) -> tuple[RecoveryRecord, ...]:
        head_revision = get_main_revision(self._paths)
        recovered: list[RecoveryRecord] = []
        for operation in list_interrupted_operations(self._connection):
            worktree_path = self._paths.worktrees_dir / operation.op_id
            worktree_revision = resolve_worktree_revision(worktree_path)
            if worktree_revision == head_revision:
                updated = mark_operation_succeeded(
                    self._connection,
                    operation.op_id,
                    result_revision=head_revision,
                    result={
                        "changed_paths": list(
                            _recover_changed_paths(
                                self._paths,
                                base_revision=operation.base_revision,
                                head_revision=head_revision,
                            )
                        )
                    },
                )
                recovered.append(
                    RecoveryRecord(
                        op_id=updated.op_id,
                        state=updated.state,
                        classification="published",
                        revision=head_revision,
                    )
                )
            elif operation.base_revision is not None and operation.base_revision != head_revision:
                updated = mark_operation_conflict(
                    self._connection,
                    operation.op_id,
                    error_message="interrupted operation is stale after startup recovery",
                )
                recovered.append(
                    RecoveryRecord(
                        op_id=updated.op_id,
                        state=updated.state,
                        classification="conflict",
                        revision=None,
                    )
                )
            else:
                recovered.append(
                    RecoveryRecord(
                        op_id=operation.op_id,
                        state=operation.state,
                        classification="retryable",
                        revision=worktree_revision,
                    )
                )
            remove_operation_worktree(self._paths, operation.op_id)
        materialized = materialize_current_checkout(self._paths, revision=head_revision)
        if self._derived_update is not None:
            self._derived_update(materialized.path, head_revision, tuple())
        return tuple(recovered)


def _recover_changed_paths(
    paths: GitRepositoryPaths, *, base_revision: str | None, head_revision: str
) -> tuple[str, ...]:
    if base_revision is None:
        return ()
    return diff_main_paths(paths, base_revision=base_revision, end_revision=head_revision)

File: go/repository/transactions.go
package repository

import (
	"context"
	"errors"
	"fmt"
	"path/filepath"
	"sort"
	"sync"
	"time"

	"github.com/rcarmo/memento/go/control"
)

type CheckpointError struct{ Name string }

func (e *CheckpointError) Error() string { return e.Name }

type TransactionConflictError struct{ Message string }

func (e *TransactionConflictError) Error() string { return e.Message }

// MutationError supplies the source diagnostic class at a callback boundary.
// Generic Go errors otherwise use their concrete Go type name.
type MutationError struct{ Class, Message string }

func (e *MutationError) Error() string { return e.Message }

type TransactionRequest struct {
	Operation                                                control.OperationRequest
	ExpectedRevision, CommitMessage, AuthorName, AuthorEmail string
}
type TransactionResult struct {
	Operation                    control.OperationRecord
	BaseRevision, ResultRevision string
	ChangedPaths                 []string
	MaterializedPath             string
	Replayed                     bool
}
type RecoveryRecord struct {
	OpID           string                 `json:"op_id"`
	State          control.OperationState `json:"state"`
	Classification string                 `json:"classification"`
	Revision       *string                `json:"revision"`
}
type MutationCallback func(context.Context, string) ([]string, error)
type DerivedUpdateCallback func(context.Context, string, string, []string) error

type TransactionManager struct {
	Operations    control.Operations
	Paths         GitRepositoryPaths
	Checkpoint    func(string) error
	DerivedUpdate DerivedUpdateCallback
	Now           func() time.Time
}

var transactionLocks = struct {
	sync.Mutex
	values map[string]*sync.Mutex
}{values: map[string]*sync.Mutex{}}

func transactionLock(paths GitRepositoryPaths) (*sync.Mutex, error) {
	return transactionLockAt(paths, filepath.Abs)
}
func transactionLockAt(paths GitRepositoryPaths, abs func(string) (string, error)) (*sync.Mutex, error) {
	absolute, err := abs(paths.BareDir)
	if err != nil {
		return nil, err
	}
	resolved, err := filepath.EvalSymlinks(absolute)
	if err == nil {
		absolute = resolved
	}
	transactionLocks.Lock()
	defer transactionLocks.Unlock()
	lock := transactionLocks.values[absolute]
	if lock == nil {
		lock = &sync.Mutex{}
		transactionLocks.values[absolute] = lock
	}
	return lock, nil
}

// WithTransactionLock serialises service control mutations with Apply/Recover
// for this repository. The caller must hold WriterLease across processes. Like
// Apply, the callback must not recursively acquire this non-reentrant lock.
func WithTransactionLock(ctx context.Context, paths GitRepositoryPaths, fn func() error) error {
	return withTransactionLock(ctx, paths, fn, transactionLock)
}
func withTransactionLock(ctx context.Context, paths GitRepositoryPaths, fn func() error, getLock func(GitRepositoryPaths) (*sync.Mutex, error)) error {
	lock, err := getLock(paths)
	if err != nil {
		return err
	}
	lock.Lock()
	defer lock.Unlock()
	if err = ctx.Err(); err != nil {
		return err
	}
	return fn()
}

func (m *TransactionManager) hit(name string) error {
	if m.Checkpoint != nil {
		return m.Checkpoint(name)
	}
	return nil
}
func (m *TransactionManager) now() time.Time {
	if m.Now != nil {
		return m.Now()
	}
	return time.Now()
}

// Apply serialises one repository in-process; the caller must hold WriterLease
// for cross-process exclusion. Callbacks must not recursively call Apply: unlike
// Python's RLock, Go mutex ownership is not associated with a goroutine.
func (m *TransactionManager) Apply(ctx context.Context, request TransactionRequest, mutate MutationCallback) (TransactionResult, error) {
	return m.applyWithLock(ctx, request, mutate, transactionLock)
}
func (m *TransactionManager) applyWithLock(ctx context.Context, request TransactionRequest, mutate MutationCallback, getLock func(GitRepositoryPaths) (*sync.Mutex, error)) (TransactionResult, error) {
	lock, err := getLock(m.Paths)
	if err != nil {
		return TransactionResult{}, err
	}
	lock.Lock()
	defer lock.Unlock()
	if err = ctx.Err(); err != nil {
		return TransactionResult{}, err
	}
	return m.ApplyUnderLock(ctx, request, mutate)
}

// ApplyUnderLock composes service policy checks and publication under one
// caller-held WithTransactionLock. The caller must also hold WriterLease.
// Calling Apply instead would recursively acquire the non-reentrant Go mutex.
func (m *TransactionManager) ApplyUnderLock(ctx context.Context, request TransactionRequest, mutate MutationCallback) (TransactionResult, error) {
	return m.apply(ctx, request, mutate, defaultTransactionGit())
}

type transactionGit struct {
	main        func(GitRepositoryPaths) (string, error)
	create      func(context.Context, GitRepositoryPaths, string, string) (Worktree, error)
	commit      func(context.Context, Worktree, []string, string, GitCommitIdentity, GitCommitIdentity) (StagedCommit, error)
	staged      func(string) ([]string, error)
	publish     func(GitRepositoryPaths, string, string) (bool, error)
	materialize func(context.Context, GitRepositoryPaths, string) (MaterializedCheckout, error)
	remove      func(GitRepositoryPaths, string) error
	resolve     func(string) (string, error)
	diff        func(context.Context, GitRepositoryPaths, string, string) ([]string, error)
}

func defaultTransactionGit() transactionGit {
	return transactionGit{GetMainRevision, CreateOperationWorktree, CommitExactPaths, ExactStagedPaths, PublishMainCompareAndSwap, MaterializeCurrentCheckout, RemoveOperationWorktree, ResolveWorktreeRevision, DiffMainPaths}
}
func (m *TransactionManager) apply(ctx context.Context, request TransactionRequest, mutate MutationCallback, git transactionGit) (result TransactionResult, err error) {
	operation, err := m.Operations.Create(ctx, request.Operation)
	if err != nil {
		return result, err
	}
	operationID := operation.OpID
	if operation.State == control.Succeeded && operation.ResultRevision != nil {
		payload, err := operation.ReplayPayload()
		if err != nil {
			return result, err
		}
		changed := []string{}
		if paths, ok := payload["changed_paths"].([]any); ok {
			for _, path := range paths {
				value, ok := path.(string)
				if !ok {
					return result, &GitError{Message: "non-string replay path is not supported"}
				}
				changed = append(changed, value)
			}
		}
		base := request.ExpectedRevision
		if operation.BaseRevision != nil && *operation.BaseRevision != "" {
			base = *operation.BaseRevision
		}
		return TransactionResult{Operation: operation, BaseRevision: base, ResultRevision: *operation.ResultRevision, ChangedPaths: changed, MaterializedPath: m.Paths.CurrentDir, Replayed: true}, nil
	}
	defer func() {
		if recovered := recover(); recovered != nil {
			err = &MutationError{Class: "RuntimeError", Message: fmt.Sprintf("transaction callback panic: %v", recovered)}
		}
		var checkpoint *CheckpointError
		var conflict *TransactionConflictError
		var uncertain *PublishOutcomeError
		if errors.As(err, &checkpoint) {
			return
		}
		// Source checkpoint failures retain worktrees. Go adds the same retention
		// for an explicitly indeterminate ref durability result, so recovery can
		// inspect the original detached HEAD instead of losing publication evidence.
		if errors.As(err, &uncertain) {
			return
		}
		if err != nil && !errors.As(err, &conflict) {
			class := fmt.Sprintf("%T", err)
			var mutation *MutationError
			if errors.As(err, &mutation) {
				class = mutation.Class
			}
			if _, updateErr := m.Operations.MarkFailed(context.WithoutCancel(ctx), operationID, class, err.Error()); updateErr != nil {
				err = updateErr
			}
		}
		if cleanupErr := git.remove(m.Paths, operationID); cleanupErr != nil {
			err = cleanupErr
		}
	}()
	if err = m.hit("operation_inserted"); err != nil {
		return result, err
	}
	base, err := git.main(m.Paths)
	if err != nil {
		return result, err
	}
	if request.ExpectedRevision != base {
		message := "expected revision " + request.ExpectedRevision + " does not match " + base
		if _, err = m.Operations.MarkConflict(ctx, operation.OpID, message); err != nil {
			return result, err
		}
		return result, &TransactionConflictError{Message: message}
	}
	operation, err = m.Operations.MarkRunning(ctx, operation.OpID, base)
	if err != nil {
		return result, err
	}
	worktree, err := git.create(ctx, m.Paths, operation.OpID, base)
	if err != nil {
		return result, err
	}
	if err = m.hit("worktree_created"); err != nil {
		return result, err
	}
	changed, err := mutate(ctx, worktree.Path)
	if err != nil {
		return result, err
	}
	sort.Strings(changed)
	if err = m.hit("mutation_applied"); err != nil {
		return result, err
	}
	identity := GitCommitIdentity{Name: request.AuthorName, Email: request.AuthorEmail, When: m.now()}
	commit, err := git.commit(ctx, worktree, changed, request.CommitMessage, identity, identity)
	if err != nil {
		return result, err
	}
	staged, err := git.staged(worktree.Path)
	if err != nil {
		return result, err
	}
	if len(staged) > 0 {
		return result, &MutationError{Class: "RuntimeError", Message: "staging area must be clean after commit"}
	}
	if err = m.hit("commit_created"); err != nil {
		return result, err
	}
	published, err := git.publish(m.Paths, base, commit.Revision)
	if err != nil {
		return result, err
	}
	if !published {
		message := "repository head moved before publication"
		if _, err = m.Operations.MarkConflict(ctx, operation.OpID, message); err != nil {
			return result, err
		}
		return result, &TransactionConflictError{Message: message}
	}
	if err = m.hit("publication_complete"); err != nil {
		return result, err
	}
	checkout, err := git.materialize(ctx, m.Paths, commit.Revision)
	if err != nil {
		return result, err
	}
	if err = m.hit("current_materialized"); err != nil {
		return result, err
	}
	if m.DerivedUpdate != nil {
		if err = m.DerivedUpdate(ctx, checkout.Path, commit.Revision, commit.ChangedPaths); err != nil {
			return result, err
		}
	}
	if err = m.hit("derived_updated"); err != nil {
		return result, err
	}
	operation, err = m.Operations.MarkSucceeded(ctx, operation.OpID, commit.Revision, map[string]any{"changed_paths": commit.ChangedPaths})
	if err != nil {
		return result, err
	}
	if err = m.hit("operation_completed"); err != nil {
		return result, err
	}
	return TransactionResult{Operation: operation, BaseRevision: base, ResultRevision: commit.Revision, ChangedPaths: commit.ChangedPaths, MaterializedPath: checkout.Path}, nil
}
func (m *TransactionManager) RecoverStartup(ctx context.Context) ([]RecoveryRecord, error) {
	return m.recover(ctx, defaultTransactionGit())
}
func (m *TransactionManager) recover(ctx context.Context, git transactionGit) ([]RecoveryRecord, error) {
	head, err := git.main(m.Paths)
	if err != nil {
		return nil, err
	}
	operations, err := m.Operations.Interrupted(ctx)
	if err != nil {
		return nil, err
	}
	recovered := []RecoveryRecord{}
	for _, operation := range operations {
		revision, err := git.resolve(filepath.Join(m.Paths.WorktreesDir, operation.OpID))
		if err != nil {
			return nil, err
		}
		record := RecoveryRecord{OpID: operation.OpID, State: operation.State, Classification: "retryable"}
		if revision != "" {
			record.Revision = &revision
		}
		if revision == head && operation.BaseRevision != nil && revision != *operation.BaseRevision {
			changed, err := git.diff(ctx, m.Paths, *operation.BaseRevision, head)
			if err != nil {
				return nil, err
			}
			updated, err := m.Operations.MarkSucceeded(ctx, operation.OpID, head, map[string]any{"changed_paths": changed})
			if err != nil {
				return nil, err
			}
			record.State = updated.State
			record.Classification = "published"
			record.Revision = &head
		} else if operation.BaseRevision != nil && *operation.BaseRevision != head {
			updated, err := m.Operations.MarkConflict(ctx, operation.OpID, "interrupted operation is stale after startup recovery")
			if err != nil {
				return nil, err
			}
			record.State = updated.State
			record.Classification = "conflict"
			record.Revision = nil
		}
		recovered = append(recovered, record)
		if err = git.remove(m.Paths, operation.OpID); err != nil {
			return nil, err
		}
	}
	checkout, err := git.materialize(ctx, m.Paths, head)
	if err != nil {
		return nil, err
	}
	if m.DerivedUpdate != nil {
		if err = m.DerivedUpdate(ctx, checkout.Path, head, []string{}); err != nil {
			return nil, err
		}
	}
	return recovered, nil
}
